File: Framework/L2DEyeBlink.cpp
/**
 *
 *  You can modify and use this source freely
 *  only for the development of application related Live2D.
 *
 *  (c) Live2D Inc. All rights reserved.
 */
#include "L2DEyeBlink.h"

namespace live2d
{
	namespace framework
	{
		L2DEyeBlink::L2DEyeBlink()
		{
			
			eyeState = STATE_FIRST ;
			
			blinkIntervalMsec = 4000 ;
			
			closingMotionMsec = 100 ;
			closedMotionMsec  =  50 ;
			openingMotionMsec = 150 ;
			
			closeIfZero = true ;
			
			
			eyeID_L = "PARAM_EYE_L_OPEN" ;
			eyeID_R = "PARAM_EYE_R_OPEN" ;
		}
		
		
		l2d_int64 L2DEyeBlink::calcNextBlink()
		{
			
			l2d_int64 time = live2d::UtSystem::getTimeMSec() ;
			double r = (double)rand() / RAND_MAX ;
			return time + (l2d_int64)( r*( 2*blinkIntervalMsec - 1 ) ) ;
		}
		
		
		void L2DEyeBlink::setInterval( int blinkIntervalMsec)
		{
			this->blinkIntervalMsec = blinkIntervalMsec ;
		}
		
		
		void L2DEyeBlink::setEyeMotion( int closingMotionMsec , int closedMotionMsec , int openingMotionMsec )
		{
			this->closingMotionMsec = closingMotionMsec ;
			this->closedMotionMsec = closedMotionMsec ;
			this->openingMotionMsec = openingMotionMsec ;
		}
		
		
		
		void L2DEyeBlink::setParam( live2d::ALive2DModel *model )
		{
			l2d_int64 time = live2d::UtSystem::getTimeMSec() ;
			float eyeParamValue ;
			float t = 0 ;
			
			switch( this->eyeState )
			{
				case STATE_CLOSING:
					
					t = (float)(( time - stateStartTime ) / (double)closingMotionMsec ) ;
					if( t >= 1 )
					{
						t = 1 ;
						this->eyeState = STATE_CLOSED ;
						this->stateStartTime = time ;
					}
					eyeParamValue = 1 - t ;
					break ;
				case STATE_CLOSED:
					t = (float)(( time - stateStartTime ) / (double)closedMotionMsec) ;
					if( t >= 1 )
					{
						this->eyeState = STATE_OPENING ;
						this->stateStartTime = time ;
					}
					eyeParamValue = 0 ;
					break ;
				case STATE_OPENING:
					t = (float)(( time - stateStartTime ) / (double)openingMotionMsec ) ;
					if( t >= 1 )
					{
						t = 1 ;
						this->eyeState = STATE_INTERVAL ;
						this->nextBlinkTime = calcNextBlink() ;
					}
					eyeParamValue = t ;
					break ;
				case STATE_INTERVAL:
					
					if( this->nextBlinkTime < time )
					{
						this->eyeState = STATE_CLOSING ;
						this->stateStartTime = time ;
					}
					eyeParamValue = 1 ;
					break ;
				case STATE_FIRST:
				default:
					this->eyeState = STATE_INTERVAL ;
					this->nextBlinkTime = calcNextBlink() ;
					eyeParamValue = 1 ;
					break ;
			}
			
			if( ! closeIfZero ) eyeParamValue = -eyeParamValue ;
			
			
			model->setParamFloat( eyeID_L.c_str() , eyeParamValue ) ;
			model->setParamFloat( eyeID_R.c_str() , eyeParamValue ) ;
			
		}
	}
}

File: package/live2d/v2/framework/motion/l2d_eye_blink.py
import random

from ...core import UtSystem


class L2DEyeBlink:

    def __init__(self):
        self.nextBlinkTime = None
        self.stateStartTime = None
        self.blinkIntervalMsec = None
        self.eyeState = EyeState.STATE_FIRST
        self.blinkIntervalMsec = 4000
        self.closingMotionMsec = 100
        self.closedMotionMsec = 50
        self.openingMotionMsec = 150
        self.closeIfZero = True
        self.eyeID_L = "PARAM_EYE_L_OPEN"
        self.eyeID_R = "PARAM_EYE_R_OPEN"

    def calcNextBlink(self):
        time = UtSystem.getUserTimeMSec()
        r = random.random()
        return time + r * (2 * self.blinkIntervalMsec - 1)

    def setInterval(self, blinkIntervalMsec):
        self.blinkIntervalMsec = blinkIntervalMsec

    def setEyeMotion(self, closingMotionMsec, closedMotionMsec, openingMotionMsec):
        self.closingMotionMsec = closingMotionMsec
        self.closedMotionMsec = closedMotionMsec
        self.openingMotionMsec = openingMotionMsec

    def updateParam(self, model):
        time = UtSystem.getUserTimeMSec()
        eye_param_value = 0

        _switch_2297 = self.eyeState
        if self.eyeState == EyeState.STATE_CLOSING:
            t = (time - self.stateStartTime) / self.closingMotionMsec
            if t >= 1:
                t = 1
                self.eyeState = EyeState.STATE_CLOSED
                self.stateStartTime = time

            eye_param_value = 1 - t
        elif self.eyeState == EyeState.STATE_CLOSED:
            t = (time - self.stateStartTime) / self.closedMotionMsec
            if t >= 1:
                self.eyeState = EyeState.STATE_OPENING
                self.stateStartTime = time

            eye_param_value = 0
        elif self.eyeState == EyeState.STATE_OPENING:
            t = (time - self.stateStartTime) / self.openingMotionMsec
            if t >= 1:
                t = 1
                self.eyeState = EyeState.STATE_INTERVAL
                self.nextBlinkTime = self.calcNextBlink()

            eye_param_value = t
        elif self.eyeState == EyeState.STATE_INTERVAL:
            if self.nextBlinkTime < time:
                self.eyeState = EyeState.STATE_CLOSING
                self.stateStartTime = time
            eye_param_value = 1
        else:
            self.eyeState = EyeState.STATE_INTERVAL
            self.nextBlinkTime = self.calcNextBlink()
            eye_param_value = 1
        if not self.closeIfZero:
            eye_param_value = -eye_param_value
        model.setParamFloat(self.eyeID_L, eye_param_value)
        model.setParamFloat(self.eyeID_R, eye_param_value)


class EyeState:
    STATE_FIRST = "STATE_FIRST"
    STATE_INTERVAL = "STATE_INTERVAL"
    STATE_CLOSING = "STATE_CLOSING"
    STATE_CLOSED = "STATE_CLOSED"
    STATE_OPENING = "STATE_OPENING"
